File: src/09.tree/46.98.go
package tree

import "math"

func IsValidBST(root *TreeNode) bool {
	var dfs func(node *TreeNode)
	ans := true
	pre := math.MinInt64
	dfs = func(node *TreeNode) {
		if node == nil {
			return
		}
		dfs(node.Left)
		if pre != math.MinInt64 && pre >= node.Val {
			ans = false
			return
		}
		pre = node.Val
		dfs(node.Right)
	}

	dfs(root)

	return ans
}

File: src/09.tree/46.98.ts
import { TreeNode } from "./01.100";

export function isValidBST(root: TreeNode | null): boolean {
    let pre = Infinity;
    let ans = true;
    const dfs = (node: TreeNode | null) => {
        if (!node) {
            return;
        }
        dfs(node.left);
        if (pre !== Infinity && pre >= node.val) {
            ans = false;
            return;
        }
        pre = node.val;
        dfs(node.right);
    };

    dfs(root);

    return ans;
}
